src/vs/workbench/contrib/notebook/browser/controller/editActions.ts

```diff
@@ -646,44 +650,151 @@ registerAction2(class CommentSelectedCellsAction extends NotebookMultiCellAction
 	async runWithContext(accessor: ServicesAccessor, context: INotebookCommandContext): Promise<void> {
 		const languageConfigurationService = accessor.get(ILanguageConfigurationService);
 
-		const selectedCellEditors: ICodeEditor[] = [];
-		context.selectedCells.forEach(cell => {
-			const findContext = { notebookEditor: context.notebookEditor, cell };
-			const foundEditor = findTargetCellEditor(findContext, cell);
-			if (foundEditor) {
-				selectedCellEditors.push(foundEditor);
+		context.selectedCells.forEach(async cell => {
+			const cellTextModel = cell.model;
+			let textModel: ITextModel | TextModel | undefined = cellTextModel.textModel;
+			if (!cellTextModel) {
+				return;
 			}
-		});
-
 
-		selectedCellEditors.forEach(editor => {
-			if (!editor.hasModel()) {
-				return;
+			if (!textModel) {
+				textModel = await cell.resolveTextModel();
 			}
 
-			const model = editor.getModel();
-			const commands: ICommand[] = [];
-			const modelOptions = model.getOptions();
-			const commentsOptions = editor.getOption(EditorOption.comments);
+			const cellEditorSelections = cell.getSelections();
+
 
-			const selection = editor.getSelection();
+			const textBuffer = cellTextModel.textBuffer;
 
-			commands.push(new LineCommentCommand(
+			// ! todo: need to store comment options on the model refer to https://github.com/microsoft/vscode/issues/108675
+			const cellCommentCommand = new LineCommentCommand(
 				languageConfigurationService,
-				new Selection(1, 1, model.getLineCount(), model.getLineMaxColumn(model.getLineCount())),
-				modelOptions.indentSize,
+				new Selection(1, 1, textBuffer.getLineCount(), textBuffer.getLineLength(textBuffer.getLineCount())), // comment the entire cell
+				textModel.getOptions().tabSize,
 				Type.Toggle,
-				commentsOptions.insertSpace,
-				commentsOptions.ignoreEmptyLines,
+				false, // commentsOptions.insertSpace,
+				false, // commentsOptions.ignoreEmptyLines,
 				false
-			));
+			);
+
+			// store any selections that are in the cell, allows them to be shifted by comments and preserved
+			const initialTrackedRangesIDs: string[] = [];
+			const initialTrackedRangesDirections: SelectionDirection[] = [];
+			cellEditorSelections.forEach(selection => {
+				initialTrackedRangesIDs.push(textModel._setTrackedRange(null, selection, TrackedRangeStickiness.NeverGrowsWhenTypingAtEdges));
+				initialTrackedRangesDirections.push(selection.getDirection());
+			});
 
-			editor.pushUndoStop();
-			editor.executeCommands(COMMENT_SELECTED_CELLS_ID, commands);
-			editor.pushUndoStop();
+			// duplicate range IDs
+			const tempTrackedRanges = initialTrackedRangesIDs.slice();
 
-			editor.setSelection(selection);
-		});
+			const ctx: IExecContext = {
```

Comment:
Since we are tracking the ranges ourselves via `textModel._setTrackedRange`, which is marker based, I found we don't need to build following edit operations anymore since we don't need any logic for tracking edits/selections. All we need is `CommandExecutor.executeCommands(textModel, [], [cellCommentCommand]);` to run the command on the text model.